synapse/handlers/sync.py

```diff
@@ -1832,10 +1839,25 @@ async def _generate_sync_entry_for_rooms(
 
         since_token = sync_result_builder.since_token
 
+        # If all rooms are blocked, we can skip bits of processing.
+        block_all_rooms = (
+            sync_result_builder.sync_config.filter_collection.blocks_all_rooms()
+        )
+
+        # 0. If there are no rooms to return *and* we don't care about presence
+        # or device list updates, there's nothing to do.
+        if (
+            block_all_rooms
+            and block_all_presence_data
+            and not include_device_list_updates
+        ):
+            return set(), set(), set(), set()
+
         # 1. Start by fetching all ephemeral events in rooms we've joined (if required).
         user_id = sync_result_builder.sync_config.user.to_string()
         block_all_room_ephemeral = (
-            since_token is None
+            block_all_rooms
+            or since_token is None
             and sync_result_builder.sync_config.filter_collection.blocks_all_room_ephemeral()
```

Comment:
Err, I can never remember the precedence rules. Is this 

1. `(A OR B) AND C` or
2. `A OR (B AND C)`?

(I think 2?)

I'm also not sure what the thinking is behind changing this condition. Is there ever a situation where `block_all_rooms` is True and `sync_result_builder.sync_config.filter_collection.blocks_all_room_ephemeral()` is False?

Reply:
This code got simplified so, doesn't matter.